proposals/text/0000-evaluation-haystack-2.md

```diff
@@ -0,0 +1,223 @@
+We shouldn't expect all input and output data to implement serialization methods like `to_dict` and `from_dict` like `Pipeline` and `component`s do, so we probably should find an alternative to handle serialization of types that don't. An unsafe option would be the use of `pickle`, but that's dangerous as it could lead to malicious code being executed.
+
+Given the above information we should be able to implement a single method to calculate predeterminated metrics or even custom ones.
+Known metrics could be defined as an enum to ease discoverability and documentation.
+
+```
+class Metric(Enum):
+    RECALL = "Recall"
+    MRR = "Mean Reciprocal Rank"
+    MAP = "Mean Average Precision"
+    EM = "Exact Match"
+    F1 = "F1"
+    SAS = "SemanticAnswerSimilarity"
+```
+
+The method to calculate metrics could look similar to this:
+
+```
+MetricsResult = Dict[str, Dict[str, float]]
+MetricCalculator = Callable[..., MetricResult]
+
+def calculate_metrics(self: EvaluationResult, metric: Union[Metric, MetricCalculator], **kwargs) -> MetricsResult:
+    # Verify if we're calculating a known metric
+    if metric == Metric.RECALL:
+      return self._calculate_recall(**kwargs)
+    elif metric == Metric.MRR:
+      return self._calculate_mrr(**kwargs)
+    # Other metrics...
+
+    # If it's not a known metric it must be a custom one
+    return metric(self, **kwargs)
+```
+
+This gives the users the flexibility to easily calculate metrics that we support but also use custom logic to calculate any kind of metric given the available data.
+Since users will need to save their calculated metrics to file we could create a simple `MetricResult` class that simply wraps the generated metrics dictionary, something similar:
+
+```
+class MetricResult(dict):
+    def save(self, file: Union[str, Path]):
+        # Dump info to file here
+```
+
+### Ease input specification
+
+As it can be annoying having to specify the inputs and outputs for evaluation since we need to specify both the `component` name and its input name we could make some assumption when running `Pipeline`s.
+
+An approach to this problem could be letting the user skip the `component` name specification when declaring the `Pipeline` input. This could work up until the point the user changes their `Pipeline` to have multiple inputs with the same name.
+
+So given a `Pipeline` that has single input `component` name `foo` that takes a `query` as its input we can let the user specify the `eval` input like so:
+
+```
+eval(pipe, {"query": "This is the query"})
+```
+
+If the user adds a new `component` name `bar` that also takes a `query` as input we'll make evaluation fail preventively since we cannot be sure whether both `component`s must take `query` as input and force explicit specification:
+
+```
+eval(pipe, {"foo": {"query": "This is the query"}, "bar": {"query": "This is the query"}})
+```
```

Comment:
Why is it so? Can't we assume that the two values are the same? I imagine that in most cases it will be the same value (and we could of course warn the user when this happen).

On a higher level this also seems a bit like an implementation detail to be fair.

Reply:
You could but that's dangerous. 

We're going to have tons of different type of `Pipeline`s so making this kind of assumptions feels extremely dangerous to me.

Making a correct assumption makes the user happy cause it removes a burden of knowledge, but making a wrong one creates only frustration cause the user must figure what's the issue of a kind of unrelated change.

I prefer to stir in the safe direction and avoid making any assumption to avoid making a frustrating mistake for the user.